### PR TITLE
Add support for subdirs in private-etc

### LINKED ISSUE #3991: private-etc doesn't allow subdirs
**Bug and expected behavior**
I want to allow a program access to the directory /etc/pki/ca-trust/ and below, but not to /etc/pki/tls/private.
As the private keys are really sensitvie, I want not just rely on blacklisting, but remove the private keys fully from the mount with the private-etc option.

When I try to use this in a profile file:
```
private-etc alternatives,crypto-policies,hostname,hosts,ld.so.cache,resolv.conf,localtime,ssl,pki/ca-trust
```
I get the error:
```
Error: "pki/ca-trust" is an invalid filename
```

The manpage suggests that copying subdirs of /etc is allowed as the example given there:
```
Example:
              $ firejail --private-etc=group,hostname,localtime, \
              nsswitch.conf,passwd,resolv.conf,default/motd-news
```
uses default/motd-news.

**Environment**
 - CentOS 8.3 on x86_64.
 - firejail 0.9.64.4

### PR DESCRIPTION
**Pull request by brisad**
Gosh it's been a while since I wrote C, and this solution turned out to use more pointer-manipulation that I'd hoped to. I wanted to avoid allocations, but this solution... I don't like it.

But since I already wrote it, and it seems to work, I thought I'd add a PR and you can take a look to see if it's anything to keep.

Fixes #3991

### CODE REVIEW

File: src/firejail/fs_etc.c [hunk 1]
```diff
@@ -76,6 +76,44 @@ void fs_machineid(void) {
 	}
 }
 
+// Duplicate directory structure from src to dst by creating empty directories.
+// The paths _must_ be identical after their respective prefixes.
+// When finished, dst will point to the target directory. That is, if
+// it starts out pointing to a file, it will instead be truncated so
+// that it contains the parent directory instead.
+static void build_dirs(char *src, char *dst, size_t src_prefix_len, size_t dst_prefix_len) {
+    char *p = src + src_prefix_len + 1;
```
Comment (kmk3) [inline src/firejail/fs_etc.c:85]: This file uses tabs for indentation.

Comment (brisad) [reply src/firejail/fs_etc.c:85]: Whoops, completely missed that. Thanks for the head up.

Comment (kmk3) [reply src/firejail/fs_etc.c:85]: > Whoops, completely missed that. Thanks for the heads up.

No problem.  So could you fix the indentation, amend the commit and force-push?

I'd suggest a command to do it automatically, but the proposed `build_dirs`
function is indented using 4 spaces and the changes in `duplicate` are indented
with a mix of 4 and 8 spaces.

Comment (kmk3) [reply src/firejail/fs_etc.c]: > Sure! I just pushed the update.
>
> Ah, yes, I even mixed tab length with my spaces, so I should've noticed that
> inconsistency :)

Thanks.

> Thanks for reviewing. As I wrote above, I'm not really happy with this one,
> so I'm all ears for suggestions for improvements. Don't have anything against
> rewriting it, perhaps using sprintf's for building up paths instead.

No problem.  I'm not really familiar with firejail's C code base, so based on
the output of `tig --no-merges master -- '*.h' '*.c'`, I think someone like
@smitsohu would be of more help in that regard.

**Comment by netblue30** [comment]
Merged in, thanks!

**Pull request merged**